### ISSUE TITLE
New version

### ISSUE THREAD

**Issue by GenaBitu** [showdownjs/showdown#568]
Hi, are there any plans on doing a new release? I need #521 and so am stuck on my own fork...

Thank you!

**Comment by tivie** [comment]
Soonish I hope.I'm working as often as I'm able to in the version 2.0 which should address all this. Unfortunately my free time is rather limited lately.

Hope I can release a version 2.0 alpha in the next few months.

**Comment by tivie** [comment]
@GenaBitu http://showdownjs.com/#!/blog/rewriting-the-parser-challenges-and-opportunities

**Comment by tivie** [comment]
I felt that version 2.0 should have a shinny and **proper new parser** that read
the code sequentially and was aware of context (use more code logic and less RexExp wizardry :tada:). The new parser is, 
in fact, a lot more similar to a PEG parser than a RegExp converter with in place substitutions.

Some key features include:

  - **Sequential**
  
    The new parser reads the input sequentially. When a syntax element match is found, it "stores" the element without
    changing the original string. This means that Showdown no longer relies on (nor needs) a specific order in which 
    sub parsers are invoked.

  - **Full separation of the *parsing* and *conversion* steps**
    
    Instead of making in place substitutions of the original input, the new parser creates an intermediary abstract layer,
    an object that is a *node tree* of elements, similar to the DOM Tree in the browser, which makes it easy to manipulate
    each node before outputting it as a string again, in other format.
    
    
  - **Output manipulation is extremely easy and customizable through templates**
  
    The node tree (and each individual node) can be manipulated as you see fit since it's, in practice, an agnostic 
    representation of an element. Each node has a template for each supported format: HTML and Markdown. But you can 
    even add more formats if you wish. 
    
    What is cool is that, **not only can you can manipulate the tree (add, delete or swap nodes)**, you can also **tweak 
    or completely change the output of all nodes of a type or even a specific, individual node**.
    
  - **Extensions are a lot more powerful now**
  
    You can pretty much change every aspect  
    
  - **Faster (it seems)**
    
    RegExps, specially the complex ones, are slow. So, moving away from them, and only using RegExp for quick checks,
    should speed up the parser a lot. Well, at least in theory.

**Comment by GenaBitu** [comment]
Oh yeah, I take it you're gonna take whatever output you get, transform it into an AST and then into whatever output is desired, right? That's definitely a more systematic and elegant solution then regex... 

One thing I would like to say though - for me personally (YMMV), the reason I chose showdown instead of e.g. [marked](https://github.com/markedjs/marked) was the simplicity with which I could extend the syntax, for example creating my own "commands" such as `!command[arg1=val1, arg2=val2]` even possibly with the command spanning multiple lines. (I did this using the `filter` function...)

So if you manage to maintain backwards compatibility for all the existing extensions while rewriting the parser for better performance, that would be awesome. Thanks.
